pubtools/pulplib/_impl/fake/match.py

```diff
@@ -112,7 +112,7 @@ def match_field_regex(matcher, field, obj):
 @visit(ExistsMatcher)
 def match_field_exists(_matcher, field, obj):
     value = get_field(field, obj)
-    return value is not ABSENT
+    return value not in [ABSENT, None]
```

Comment:
Not sure if it's going to be this easy.  Actually, see the TODO comment which exists later in this file about whether we need to differentiate between absent and present but None.

The issue here is that, if the value was None, you have no way of knowing if it was explicitly set to None or if it simply wasn't present.  For example, consider eng_product_id.  On a real Pulp server, it comes from field notes.eng_product.  That field could have been absent altogether, or it could have been present but with a value of null.  If it was present and null, then mongo $exists fragment would have found it.  So one can argue that exists() with the fake client should then *also* find objects where the field is None.

PulpObject classes simply don't hold the necessary info right now to differentiate between "this field is absent" and "this field is present but None", so I am not sure what's the most correct way to handle this right now.

If this change is solving some problem, it may make sense to proceed with it, as both old and new behaviors can be considered wrong from certain perspectives, so whichever works better in practice could be the better choice.

Either way, could we please get an autotest?

Reply:
I thought about this some more and became more convinced that this behavior change you want to make is correct.  With fake client, our choices seem to be one of these caveats:

1. either it's not possible to represent an absent field
2. or it's not possible to represent a present-but-null field

I implemented it originally as (1) and you're now changing it to (2).  I guess that is actually the better choice simply because absent fields are very common while present-but-null fields are uncommon.

Reply:
Agree. And this is a problem only with exists matcher which not be very prevalent.
I'll update this patch with test cases